File: android/app/src/main/java/com/bluelink/android/domain/DraftLedger.kt
package com.bluelink.android.domain

data class DraftSnapshot(val peerId: String, val text: String, val revision: Long)

/** Revisions protect later edits (including A -> B -> A) from asynchronous send/save callbacks. */
class DraftLedger {
    private val values = linkedMapOf<String, DraftSnapshot>()
    private val dirty = mutableSetOf<String>()
    private var revision = 0L
    private fun key(peerId: String) = peerId.lowercase(java.util.Locale.ROOT)

    @Synchronized fun load(stored: Map<String, String>) {
        values.clear(); dirty.clear()
        stored.forEach { (peer, text) -> values[key(peer)] = DraftSnapshot(key(peer), text, ++revision) }
    }
    @Synchronized fun get(peerId: String): DraftSnapshot =
        values[key(peerId)] ?: DraftSnapshot(key(peerId), "", 0)
    @Synchronized fun edit(peerId: String, text: String): DraftSnapshot {
        val peer = key(peerId)
        if (get(peer).text == text) return get(peer)
        return DraftSnapshot(peer, text, ++revision).also { values[peer] = it; dirty += peer }
    }
    @Synchronized fun clearAfterSend(snapshot: DraftSnapshot): Boolean {
        if (get(snapshot.peerId) != snapshot) return false
        edit(snapshot.peerId, "")
        return true
    }
    @Synchronized fun acknowledge(snapshot: DraftSnapshot) {
        if (get(snapshot.peerId) == snapshot) dirty -= key(snapshot.peerId)
    }
    @Synchronized fun pending(): List<DraftSnapshot> = dirty.mapNotNull(values::get)
    @Synchronized fun texts(): Map<String, String> = values.mapValues { it.value.text }
    @Synchronized fun clear(peerId: String? = null) {
        if (peerId != null) edit(peerId, "") else values.keys.toList().forEach { edit(it, "") }
    }
}

File: windows/BlueLink.App/Domain/DraftLedger.cs
namespace BlueLink.Domain;

internal sealed record DraftSnapshot(string PeerId, string Text, long Revision);

/// <summary>Send completion and disk writes only acknowledge the exact captured edit.</summary>
internal sealed class DraftLedger
{
    private readonly object _gate = new();
    private readonly Dictionary<string, DraftSnapshot> _values = new(StringComparer.OrdinalIgnoreCase);
    private readonly HashSet<string> _dirty = new(StringComparer.OrdinalIgnoreCase);
    private long _revision;
    internal void Load(IEnumerable<KeyValuePair<string, string>> stored)
    {
        lock (_gate)
        {
            _values.Clear(); _dirty.Clear();
            foreach (var (peer, text) in stored)
                _values[peer] = new(peer.ToLowerInvariant(), text, ++_revision);
        }
    }
    internal DraftSnapshot Get(string peer)
    {
        lock (_gate) return _values.GetValueOrDefault(peer) ?? new(peer.ToLowerInvariant(), "", 0);
    }
    internal DraftSnapshot Edit(string peer, string text)
    {
        lock (_gate)
        {
            var existing = Get(peer);
            if (existing.Text == text) return existing;
            var value = new DraftSnapshot(peer.ToLowerInvariant(), text, ++_revision);
            _values[peer] = value; _dirty.Add(peer); return value;
        }
    }
    internal bool ClearAfterSend(DraftSnapshot snapshot)
    {
        lock (_gate)
        {
            if (Get(snapshot.PeerId) != snapshot) return false;
            Edit(snapshot.PeerId, ""); return true;
        }
    }
    internal void Acknowledge(DraftSnapshot snapshot)
    {
        lock (_gate) if (Get(snapshot.PeerId) == snapshot) _dirty.Remove(snapshot.PeerId);
    }
    internal IReadOnlyList<DraftSnapshot> Pending()
    {
        lock (_gate) return _dirty.Select(peer => _values[peer]).ToArray();
    }
    internal void Clear(string? peer = null)
    {
        lock (_gate)
        {
            if (peer is not null) Edit(peer, "");
            else foreach (var id in _values.Keys.ToArray()) Edit(id, "");
        }
    }
}
